server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java

```diff
@@ -0,0 +1,425 @@
+import static org.apache.lucene.geo.GeoUtils.orient;
+
+/**
+ * A tree reusable reader for a previous serialized {@link org.elasticsearch.geometry.Geometry} using
+ * {@link TriangleTreeWriter}.
+ *
+ * This class supports checking bounding box
+ * relations against the serialized triangle tree.
+ *
+ * -----------------------------------------
+ * |   The binary format of the tree       |
+ * -----------------------------------------
+ * -----------------------------------------  --
+ * |    centroid-x-coord (4 bytes)         |    |
+ * -----------------------------------------    |
+ * |    centroid-y-coord (4 bytes)         |    |
+ * -----------------------------------------    |
+ * |    DimensionalShapeType (1 byte)      |    | Centroid-related header
+ * -----------------------------------------    |
+ * |  Sum of weights (VLong 1-8 bytes)     |    |
+ * -----------------------------------------  --
+ * |         Extent (var-encoding)         |
+ * -----------------------------------------
+ * |         Triangle Tree                 |
+ * -----------------------------------------
+ * -----------------------------------------
+ */
+public class TriangleTreeReader {
+    private final ByteArrayDataInput input;
+    private final CoordinateEncoder coordinateEncoder;
+    private final Rectangle2D rectangle2D;
+    private final Extent extent;
+    private int treeOffset;
+
+    public TriangleTreeReader(CoordinateEncoder coordinateEncoder) {
+        this.coordinateEncoder = coordinateEncoder;
+        this.rectangle2D = new Rectangle2D();
+        this.extent = new Extent();
+        this.input = new ByteArrayDataInput();
+    }
+
+    public void reset(BytesRef bytesRef) throws IOException {
+        this.input.reset(bytesRef.bytes, bytesRef.offset, bytesRef.length);
+        treeOffset = 0;
+    }
+
+    /**
+     * returns the bounding box of the geometry in the format [minX, maxX, minY, maxY].
+     */
+    public Extent getExtent() {
+        if (treeOffset == 0) {
+            getSumCentroidWeight(); // skip CENTROID_HEADER + var-long sum-weight
+            Extent.readFromCompressed(input, extent);
+            treeOffset = input.getPosition();
+        } else {
+            input.setPosition(treeOffset);
+        }
+        return extent;
+    }
+
+    /**
+     * returns the X coordinate of the centroid.
+     */
+    public double getCentroidX() {
+        input.setPosition(0);
+        return coordinateEncoder.decodeX(input.readInt());
+    }
+
+    /**
+     * returns the Y coordinate of the centroid.
+     */
+    public double getCentroidY() {
+        input.setPosition(4);
+        return coordinateEncoder.decodeY(input.readInt());
+    }
+
+    public DimensionalShapeType getDimensionalShapeType() {
+        input.setPosition(8);
+        return DimensionalShapeType.readFrom(input);
+    }
+
+    public double getSumCentroidWeight() {
+        input.setPosition(9);
+        return Double.longBitsToDouble(input.readVLong());
```

Comment:
it's almost always wasteful to encode double bits as a vLong, doing a writeLong/readLong would be better

Reply:
do you mean it would be ideal if we could keep track of the centroid weight as a `long` from the beginning?